src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs

```diff
@@ -49,6 +65,185 @@ public IEnumerable<Document> GetDocumentsByBucketFrom(DocumentsOperationContext
+    internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForDocuments(ByteStringContext context, ref TableValueReader tvr, out Slice slice)
+    {
+        return GenerateBucketAndEtagIndexKey(context, idIndex: (int)DocumentsTable.LowerId, etagIndex: (int)DocumentsTable.Etag, ref tvr, out slice);
+    }
+
+    [StorageIndexEntryKeyGenerator]
+    internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForTombstones(ByteStringContext context, ref TableValueReader tvr, out Slice slice)
+    {
+        return GenerateBucketAndEtagIndexKey(context, idIndex: (int)TombstoneTable.LowerId, etagIndex: (int)TombstoneTable.Etag, ref tvr, out slice);
+    }
+
+    internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKey(ByteStringContext context, int idIndex, int etagIndex, ref TableValueReader tvr, out Slice slice)
+    {
+        var idPtr = tvr.Read(idIndex, out var size);
+        var etag = *(long*)tvr.Read(etagIndex, out _);
+
+        return GenerateBucketAndEtagSlice(context, idPtr, size, etag, out slice);
+    }
+
+    internal static ByteStringContext.Scope ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(ByteStringContext context, int keyIndex, int etagIndex, ref TableValueReader tvr, out Slice slice)
+    {
+        var keyPtr = tvr.Read(keyIndex, out var keySize);
+
+        int sizeOfDocId = 0;
+        for (; sizeOfDocId < keySize; sizeOfDocId++)
+        {
+            if (keyPtr[sizeOfDocId] == SpecialChars.RecordSeparator)
+                break;
+        }
+
+        var etag = *(long*)tvr.Read(etagIndex, out _);
+
+        return GenerateBucketAndEtagSlice(context, keyPtr, sizeOfDocId, etag, out slice);
+    }
+
+    private static ByteStringContext.Scope GenerateBucketAndEtagSlice(ByteStringContext context, byte* idPtr, int idSize, long etag, out Slice slice)
+    {
+        var scope = context.Allocate(sizeof(long) + sizeof(int), out var buffer);
+
+        var bucket = ShardHelper.GetBucket(idPtr, idSize);
+
+        *(int*)buffer.Ptr = Bits.SwapBytes(bucket);
+        *(long*)(buffer.Ptr + sizeof(int)) = etag;
+
+        slice = new Slice(buffer);
+        return scope;
+    }
+
+    internal static void UpdateBucketStatsForDocument(Transaction tx, Slice key, long oldSize, long newSize)
+    {
+        int numOfDocsChanged = 0;
+        if (oldSize == 0)
+            numOfDocsChanged = 1;
+        else if (newSize == 0)
+            numOfDocsChanged = -1;
+
+        UpdateBucketStatsInternal(tx, key, oldSize, newSize, numOfDocsChanged);
+    }
+
+    internal static void UpdateBucketStats(Transaction tx, Slice key, long oldSize, long newSize)
+    {
+        UpdateBucketStatsInternal(tx, key, oldSize, newSize, numOfDocsChanged: 0);
+    }
+
+    private static void UpdateBucketStatsInternal(Transaction tx, Slice key, long oldSize, long newSize, int numOfDocsChanged)
+    {
+        if (tx.Owner is not ShardedDocumentDatabase documentDatabase)
+            return;
```

Comment:
is there a chance we can have a regular DocumentDatabase here in a normal flow? Maybe we should throw instead?

Reply:
yes, in current impl. we can have a regular DocumentDatabase here: https://github.com/ravendb/ravendb/blob/fec27c41265f53f714f2851f5bd97f4796d0812d/src/Raven.Server/Documents/DocumentsTransaction.cs#L28-#L36
should I change it such that `tx.Owner` will be set only for sharded database?